File: Gaussian_Gibbs.cpp
#include <iostream>
#include <cmath>
#include<fstream>
const int niter=10000;
/******************************************************************/
/*** Gaussian Random Number Generator with Box Muller Algorithm ***/
/******************************************************************/
int BoxMuller(double& p, double& q){
  
  double pi=2e0*asin(1e0);
  //uniform random numbers between 0 and 1
  double r = (double)rand()/RAND_MAX;
  double s = (double)rand()/RAND_MAX;
  //Gaussian random numbers, 
  //with weights proportional to e^{-p^2/2} and e^{-q^2/2}
  p=sqrt(-2e0*log(r))*sin(2e0*pi*s);
  q=sqrt(-2e0*log(r))*cos(2e0*pi*s);
  
  return 0;
}
int main()
{
  double A[3][3];
  A[0][0]=1e0;A[1][1]=2e0;A[2][2]=2e0;
  A[0][1]=1e0;A[0][2]=1e0;A[1][2]=1e0;
  A[1][0]=A[0][1];A[2][0]=A[0][2];A[2][1]=A[1][2];

  srand((unsigned)time(NULL));
  /*********************************/
  /* Set the initial configuration */
  /*********************************/
  double x=0e0;double y=0e0;double z=0e0;
  /*****************/
  /*** Main part ***/
  /*****************/
  for(int iter=0; iter!=niter; iter++){
    double sigma,mu;
    double r1,r2;    
    //update x
    sigma=1e0/sqrt(A[0][0]);
    mu=-A[0][1]/A[0][0]*y-A[0][2]/A[0][0]*z;
    BoxMuller(r1,r2);
    x=sigma*r1+mu;
    //update y
    sigma=1e0/sqrt(A[1][1]);
    mu=-A[1][0]/A[1][1]*x-A[1][2]/A[1][1]*z;
    BoxMuller(r1,r2);
    y=sigma*r1+mu;
    //update z
    sigma=1e0/sqrt(A[2][2]);
    mu=-A[2][0]/A[2][2]*x-A[2][1]/A[2][2]*y;
    BoxMuller(r1,r2);
    z=sigma*r1+mu;
    // output x,y,z
    if((iter+1)%10==0){
      std::cout << std::fixed << std::setprecision(6) 
        << x << "   " 
        << y << "   " 
        << z << std::endl;
    }
  } 
  
  return 0;
}
  

File: Gaussian_Gibbs.py
import numpy as np
from matplotlib import pyplot as plt
# seed of random numbers is the system time by default
# Numpy has Gaussian random number generator, so we don't use Box-Muller

NITER = 10000
A = [[1,0,1.0,1.0],
     [1.0,2.0,1.0],
     [1.0,1.0,2.0]]
A = np.array(A)
#####################################
### Set the initial configuration ###
#####################################
x = 0
y = 0
z = 0
#################
### Main part ###
#################

# for plot
data1=[]
data2=[]
data3=[]

for iter in range(NITER):
    #update x
    sigma = 1.0/np.sqrt(A[0][0])
    mu = -A[0][1]/A[0][0]*y-A[0][2]/A[0][0]*z
    r = np.random.randn()
    x = sigma*r+mu
    #update y
    sigma = 1.0/np.sqrt(A[1][1])
    mu = -A[1][0]/A[1][1]*x-A[1][2]/A[1][1]*z
    r = np.random.randn()
    y = sigma*r+mu
    #update z
    sigma = 1.0/np.sqrt(A[2][2])
    mu = -A[2][0]/A[2][2]*x-A[2][1]/A[2][2]*y
    r = np.random.randn()
    z = sigma*r+mu
    #output x,y,z
    print(x,y,z)
    data1.append(x)
    data2.append(y)
    data3.append(z)

############
### plot ###
############
plt.figure(figsize=(15,5))
plt.axes().set_aspect('equal')
##
ax1 = plt.subplot(1, 3, 1)
ax1.scatter(data1,data2,c="black", s=20, marker="+")
ax1.set_xlabel("x")
ax1.set_ylabel("y")
ax1.set_xlim([-8,8])
ax1.set_ylim([-8,8])
ax1.set_title("x-y")
ax1.grid(linestyle = "--")
##
ax2 = plt.subplot(1, 3, 2)
ax2.scatter(data1,data3,c="blue", s=20, marker="+")
ax2.set_xlim([-8,8])
ax2.set_ylim([-8,8])
ax2.set_xlabel("x")
ax2.set_ylabel("z")
ax2.set_title("x-z")
ax2.grid(linestyle = "--")
###
ax3 = plt.subplot(1, 3, 3)
ax3.scatter(data2,data3,c="red", s=20, marker="+")
ax3.set_xlim([-8,8])
ax3.set_ylim([-8,8])
ax3.set_xlabel("y")
ax3.set_ylabel("z")
ax3.set_title("y-z")
ax3.grid(linestyle = "--")
plt.show()

